conda/plan.py

```diff
@@ -231,87 +234,88 @@ def plan_from_actions(actions):
     return res
 
 
-def extracted_where(dist):
-    for pkgs_dir in config.pkgs_dirs:
-        if install.is_extracted(pkgs_dir, dist):
-            return pkgs_dir
-    return None
-
-
-def ensure_linked_actions(dists, prefix):
+def linked_actions(dists, prefix, force, index=None):
     actions = defaultdict(list)
     actions[inst.PREFIX] = prefix
+    actions['op_order'] = (inst.RM_FETCHED, inst.FETCH, inst.RM_EXTRACTED,
+                           inst.EXTRACT, inst.UNLINK, inst.LINK)
     for dist in dists:
-        if install.is_linked(prefix, dist):
+        fetched_in = install.is_fetched(dist)
+        extracted_in = install.is_extracted(dist)
+
+        if fetched_in and index is not None:
+            # Test the MD5, and possibly re-fetch
+            try:
+                if md5_file(fetched_in) != index[dist + '.tar.bz2']['md5']:
+                    # RM_FETCHED now removes the extracted data too
+                    actions[inst.RM_FETCHED].append(dist)
+                    # Re-fetch, re-extract, re-link
+                    fetched_in = extracted_in = None
+                    force = True
+            except KeyError:
+                sys.stderr.write('Warning: cannot lookup MD5 of: %s' % fn)
+
+        if not force and install.is_linked(prefix, dist):
             continue
 
-        extracted_in = extracted_where(dist)
-        if extracted_in:
+        if extracted_in and force:
+            # Always re-extract in the force case
+            actions[inst.RM_EXTRACTED].append(dist)
+            extracted_in = None
+
+        # Otherwise we need to extract, and possibly fetch
+        if not fetched_in:
+            # If there is a cache conflict, clean it up
+            fetched_in, conflict = install.find_new_location(dist)
+            if conflict is not None:
+                actions[inst.RM_FETCHED].append(conflict)
+            actions[inst.FETCH].append(dist)
+
+        if not extracted_in:
+            actions[inst.EXTRACT].append(dist)
+
+        fetched_dist = fetched_in[:-8]
+        fetched_dir = dirname(fetched_in)
+
```

Comment:
OK, it looks like `fetched_in` here is expecting to see the full `.tar.bz2` filename, and not the directory.